fleet provisioning: could not rewrite old cert/key

**Describe the bug**
Provisioning can't be completed on next attempts when first attempt failed after key and certificate files has been stored for example when thing was not created.
Key and certificate files are created with 0600 permissions on Linux. When first attempt failed after files were created next attempts can't overwrite these files.


**To Reproduce**
Steps to reproduce:
1. Broke provision  at thing creation step for example: provide invalid provisioning template or ban device by id from provisioning lambda on AWS IoT Core..
2. Start first attempt of fleet provisioning by claim.
3. Ensure certificate and key files have been created but device is not actually provisioned with some exception during thing creation phase.
4. Fix backend problem which prevents thing creation.
5. Start second attempt of fleet provisioning by claim.
6. Check the greengrass.log file 

**Expected behavior**
 Next attempts after failed can be successful.

**Actual behavior**
Seems problem becomes persistent. 

Possible solution:
1. Drop existing key and certificate files on each new attempt.
2. Reuse existing key and certificate files and restart attempt from thing creation step (skip certificate/key creation).

**Environment**
 - OS: expecting the same on Linux (ask me about actual OS)
 - JDK version: 11
 - Nucleus version: 2.8.1
 - AWS Fleet Provisioning plugin version: v1.1.0 (e8b986af4e59273d54a0f15fd0b4d55e7a9f3d96)

**Additional context**
Temporary issue for example network connectivity can cause persistent problem when device could not do provisioning at all.

130|device:/Nucleus/files/greengrass/v2/logs # ls -la ..
total 112
drwxr-xr-x 12 ggs ggs 4096 2022-11-29 14:12 .
drwx------  5 ggs ggs 4096 2022-11-29 14:08 ..
drwx------  3 ggs ggs 4096 2022-11-29 14:12 alts
drwxr-xr-x  2 ggs ggs 4096 2022-11-29 14:12 bin
drwxr-xr-x  2 ggs ggs 4096 2022-11-29 14:12 cli_ipc_info
drwx------  2 ggs ggs 4096 2022-11-29 14:13 config
drwx------  2 ggs ggs 4096 2022-11-29 14:12 deployments
drwx------  2 ggs ggs 4096 2022-11-29 14:12 logs
drwxr-xr-x  5 ggs ggs 4096 2022-11-29 14:12 packages
drwx------  4 ggs ggs 4096 2022-11-29 14:12 plugins
-r--------  1 ggs ggs 1679 2022-11-29 14:12 privKey.key
drwx------  2 ggs ggs 4096 2022-11-29 14:12 telemetry
-r--------  1 ggs ggs 1224 2022-11-29 14:12 thingCert.crt
drwxr-xr-x  2 ggs ggs 4096 2022-11-29 14:12 work


2022-11-29T14:21:49.076Z [INFO] (pool-23-thread-1) com.aws.greengrass.IotIdentityHelper: Published to CreateKeysAndCertificate. {}
2022-11-29T14:21:49.549Z [ERROR] (pool-23-thread-1) com.aws.greengrass.FleetProvisioningByClaimPlugin: Caught exception while writing certificate and private key to file. {}
2022-11-29T14:21:49.554Z [ERROR] (pool-23-thread-1) com.aws.greengrass.lifecyclemanager.KernelLifecycle: Caught exception while running provisioning plugin. Moving on to run Greengrass without provisioning. {}
java.lang.RuntimeException: java.nio.file.AccessDeniedException: /data/data/aws.greengrass.Nucleus/files/greengrass/v2/thingCert.crt
        at com.aws.greengrass.FleetProvisioningByClaimPlugin.writeCertificateAndKeyToPath(FleetProvisioningByClaimPlugin.java:263)
        at com.aws.greengrass.FleetProvisioningByClaimPlugin.updateIdentityConfiguration(FleetProvisioningByClaimPlugin.java:155)
        at com.aws.greengrass.lifecyclemanager.KernelLifecycle.lambda$executeProvisioningPlugin$1(KernelLifecycle.java:211)
        at com.aws.greengrass.lifecyclemanager.KernelLifecycle$$ExternalSyntheticLambda2.apply(Unknown Source:4)
        at com.aws.greengrass.util.RetryUtils.runWithRetry(RetryUtils.java:50)
        at com.aws.greengrass.lifecyclemanager.KernelLifecycle.lambda$executeProvisioningPlugin$2$com-aws-greengrass-lifecyclemanager-KernelLifecycle(KernelLifecycle.java:210)
        at com.aws.greengrass.lifecyclemanager.KernelLifecycle$$ExternalSyntheticLambda3.run(Unknown Source:6)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:920)
Caused by: java.nio.file.AccessDeniedException: /data/data/aws.greengrass.Nucleus/files/greengrass/v2/thingCert.crt
        at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
        at java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:434)
        at java.nio.file.Files.newOutputStream(Files.java:216)
        at java.nio.file.Files.write(Files.java:3292)
        at com.aws.greengrass.FleetProvisioningByClaimPlugin.writeCertificateAndKeyToPath(FleetProvisioningByClaimPlugin.java:252)
        ... 9 more

--- a/src/main/java/com/aws/greengrass/FleetProvisioningByClaimPlugin.java
+++ b/src/main/java/com/aws/greengrass/FleetProvisioningByClaimPlugin.java
@@ -274,14 +274,14 @@ public class FleetProvisioningByClaimPlugin implements DeviceIdentityInterface {
                 Files.createFile(certPath);
             }
             Files.write(certPath, response.certificatePem.getBytes(StandardCharsets.UTF_8));
-            Platform.getInstance().setPermissions(FileSystemPermission.builder().ownerRead(true).build(), certPath);
+            Platform.getInstance().setPermissions(FileSystemPermission.builder().ownerRead(true).ownerWrite(true).build(), certPath);
 
             Path keyPath = Paths.get(rootPath, PRIVATE_KEY_PATH_RELATIVE_TO_ROOT);
             if (Files.notExists(keyPath)) {
                 Files.createFile(keyPath);
             }
             Files.write(keyPath, response.privateKey.getBytes(StandardCharsets.UTF_8));
-            Platform.getInstance().setPermissions(FileSystemPermission.builder().ownerRead(true).build(), keyPath);
+            Platform.getInstance().setPermissions(FileSystemPermission.builder().ownerRead(true).ownerWrite(true).build(), keyPath);
         } catch (IOException e) {
             logger.atError().log("Caught exception while writing certificate and private key to file");
             throw new DeviceProvisioningRuntimeException("Failed to write certificate and private key", e);
@@ -293,8 +293,7 @@ public class FleetProvisioningByClaimPlugin implements DeviceIdentityInterface {
             Path certPath = Paths.get(rootPath, DEVICE_CERTIFICATE_PATH_RELATIVE_TO_ROOT);
             Files.createFile(certPath);
             Files.write(certPath, response.certificatePem.getBytes(StandardCharsets.UTF_8));
-            Platform.getInstance().setPermissions(FileSystemPermission.builder()
-                    .ownerRead(true).ownerWrite(true).ownerExecute(true).build(), certPath);
+            Platform.getInstance().setPermissions(FileSystemPermission.builder().ownerRead(true).ownerWrite(true).build(), certPath);
         } catch (IOException e) {
             logger.atError().log("Caught exception while writing certificate to file", e);
             throw new DeviceProvisioningRuntimeException("Failed to write certificate", e);

--- a/src/main/java/com/aws/greengrass/FleetProvisioningByClaimPlugin.java
+++ b/src/main/java/com/aws/greengrass/FleetProvisioningByClaimPlugin.java
@@ -274,14 +274,16 @@ public class FleetProvisioningByClaimPlugin implements DeviceIdentityInterface {
                 Files.createFile(certPath);
             }
             Files.write(certPath, response.certificatePem.getBytes(StandardCharsets.UTF_8));
-            Platform.getInstance().setPermissions(FileSystemPermission.builder().ownerRead(true).ownerWrite(true).build(), certPath);
+            Platform.getInstance().setPermissions(FileSystemPermission.builder().ownerRead(true)
+                .ownerWrite(true).build(), certPath);
 
             Path keyPath = Paths.get(rootPath, PRIVATE_KEY_PATH_RELATIVE_TO_ROOT);
             if (Files.notExists(keyPath)) {
                 Files.createFile(keyPath);
             }
             Files.write(keyPath, response.privateKey.getBytes(StandardCharsets.UTF_8));
-            Platform.getInstance().setPermissions(FileSystemPermission.builder().ownerRead(true).ownerWrite(true).build(), keyPath);
+            Platform.getInstance().setPermissions(FileSystemPermission.builder().ownerRead(true)
+                .ownerWrite(true).build(), keyPath);
         } catch (IOException e) {
             logger.atError().log("Caught exception while writing certificate and private key to file");
             throw new DeviceProvisioningRuntimeException("Failed to write certificate and private key", e);
@@ -293,7 +295,8 @@ public class FleetProvisioningByClaimPlugin implements DeviceIdentityInterface {
             Path certPath = Paths.get(rootPath, DEVICE_CERTIFICATE_PATH_RELATIVE_TO_ROOT);
             Files.createFile(certPath);
             Files.write(certPath, response.certificatePem.getBytes(StandardCharsets.UTF_8));
-            Platform.getInstance().setPermissions(FileSystemPermission.builder().ownerRead(true).ownerWrite(true).build(), certPath);
+            Platform.getInstance().setPermissions(FileSystemPermission.builder().ownerRead(true)
+                .ownerWrite(true).build(), certPath);
         } catch (IOException e) {
             logger.atError().log("Caught exception while writing certificate to file", e);
             throw new DeviceProvisioningRuntimeException("Failed to write certificate", e);

